src/main/java/com/puppycrawl/tools/checkstyle/api/FullIdent.java

```diff
@@ -77,37 +77,90 @@ public static FullIdent createFullIdent(DetailAST ast) {
      */
     private static void extractFullIdent(FullIdent full, DetailAST ast) {
         if (ast != null) {
-            final DetailAST nextSibling = ast.getNextSibling();
-
-            // Here we want type declaration, but not initialization
-            final boolean isArrayTypeDeclarationStart = nextSibling != null
-                    && (nextSibling.getType() == TokenTypes.ARRAY_DECLARATOR
-                        || nextSibling.getType() == TokenTypes.ANNOTATIONS)
-                    && isArrayTypeDeclaration(nextSibling);
-
             final int typeOfAst = ast.getType();
-            if (typeOfAst == TokenTypes.LITERAL_NEW
-                    && ast.hasChildren()) {
-                final DetailAST firstChild = ast.getFirstChild();
-                extractFullIdent(full, firstChild);
-            }
-            else if (typeOfAst == TokenTypes.DOT) {
-                final DetailAST firstChild = ast.getFirstChild();
-                extractFullIdent(full, firstChild);
-                full.append(".");
-                extractFullIdent(full, firstChild.getNextSibling());
-                appendBrackets(full, ast);
-            }
-            else if (isArrayTypeDeclarationStart) {
-                full.append(ast);
-                appendBrackets(full, ast);
-            }
-            else if (typeOfAst != TokenTypes.ANNOTATIONS) {
-                full.append(ast);
+            switch (typeOfAst) {
+                case TokenTypes.DOT:
+                    extractFullIdentFromDotAst(full, ast);
+                    break;
+                case TokenTypes.TYPECAST:
+                    extractFullIdentFromTypecastAst(full, ast);
+                    break;
+                case TokenTypes.LITERAL_NEW:
+                    if (ast.hasChildren()) {
+                        extractFullIdent(full, ast.getFirstChild());
+                    }
+                    else {
+                        full.append(ast);
+                    }
+                    break;
+                case TokenTypes.METHOD_CALL:
+                case TokenTypes.INDEX_OP:
+                    extractFullIdent(full, ast.getFirstChild());
+                    break;
+                default:
+                    if (isArrayTypeDeclarationStart(ast.getNextSibling())) {
+                        full.append(ast);
+                        appendBrackets(full, ast);
+                    }
+                    else if (typeOfAst != TokenTypes.ANNOTATIONS) {
+                        full.append(ast);
+                    }
+                    break;
             }
         }
     }
 
+    /**
+     * Extract full ident from dot ast.
+     *
+     * @param full the FullIdent to add to
+     * @param dotAst ast node of type {@link TokenTypes#DOT}
+     */
+    private static void extractFullIdentFromDotAst(FullIdent full, DetailAST dotAst) {
+        DetailAST firstChildToExtract = dotAst.getFirstChild();
```

Comment:
Yes, this would be required too. All discussion at https://github.com/checkstyle/checkstyle/pull/11672 and surrounding issues indicate that changing FullIdent will break some classes even if the implementation of FullIdent is correct (which is this one; the master codebase is buggy).

I would also deprecate the usage of FullIdent as it is buggy, and fixing it would require fixing VariableDeclarationDistanceUsage. 

Either `VariableDeclarationDistanceUsage` should have a separate implementation for FullIdent (the current buggy one), which will allow us to fix the main one. What do you suggest?